drivers/net/mvpp2/mrvl_ptp.c

```diff
@@ -0,0 +1,346 @@
+			(void *)priv);
+	if (ret == 0) {
+		priv->scheduled_get_time_alarm;
+	}
+	return ret;
+}
+
+void mvpp2_cancel_phc_alarm(struct mrvl_priv *priv)
+{
+	if (priv->scheduled_get_time_alarm) {
+		MRVL_LOG(DEBUG, "### %u-%u: cancel get time from PHC alarm",
+				priv->ppio->pp2_id, priv->ppio->port_id);
+		rte_eal_alarm_cancel(mvpp2_phc_get_current_time, (void *)priv);
+		priv->scheduled_get_time_alarm = false;
+	}
+}
+
+/* PTPAction */
+enum mvpp22_ptp_action {
+	MVPP22_PTP_ACTION_NONE = 0,
+	MVPP22_PTP_ACTION_FORWARD = 1,
+	MVPP22_PTP_ACTION_CAPTURE = 3,
+	/* The following have not been verified */
+	MVPP22_PTP_ACTION_ADDTIME = 4,
+	MVPP22_PTP_ACTION_ADDCORRECTEDTIME = 5,
+	MVPP22_PTP_ACTION_CAPTUREADDTIME = 6,
+	MVPP22_PTP_ACTION_CAPTUREADDCORRECTEDTIME = 7,
+	MVPP22_PTP_ACTION_ADDINGRESSTIME = 8,
+	MVPP22_PTP_ACTION_CAPTUREADDINGRESSTIME = 9,
+	MVPP22_PTP_ACTION_CAPTUREINGRESSTIME = 10,
+};
+
+/* PTPPacketFormat */
+enum mvpp22_ptp_packet_format {
+	MVPP22_PTP_PKT_FMT_PTPV2 = 0,
+	MVPP22_PTP_PKT_FMT_PTPV1 = 1,
+	MVPP22_PTP_PKT_FMT_Y1731 = 2,
+	MVPP22_PTP_PKT_FMT_NTPTS = 3,
+	MVPP22_PTP_PKT_FMT_NTPRX = 4,
+	MVPP22_PTP_PKT_FMT_NTPTX = 5,
+	MVPP22_PTP_PKT_FMT_TWAMP = 6,
+};
+
+#define MVPP22_PTP_ACTION(x)		(((x) & 15) << 0)
+#define MVPP22_PTP_PACKETFORMAT(x)	(((x) & 7) << 4)
+#define MVPP22_PTP_MACTIMESTAMPINGEN	BIT(11)
+#define MVPP22_PTP_TIMESTAMPENTRYID(x)	(((x) & 31) << 12)
+#define MVPP22_PTP_TIMESTAMPQUEUESELECT	BIT(18)
+
+#define MVPP22_PTP_DESC_MASK_LOW  0xfff
+#define PTP_CLASS_V1    0x01 /* protocol version 1 */
+#define PTP_CLASS_VMASK 0x0f /* max protocol version is 15 */
+
+bool mvpp2_tx_hw_tstamp(struct mrvl_priv *priv,
+			       struct pp2_ppio_desc *tx_desc,
+			       struct rte_mbuf *mbuf)
+{
+	struct mvpp2_hwtstamp_queue *queue;
+	unsigned int mtype, type, i;
+	struct ptp_header *hdr;
+	u64 ptpdesc;
+
+#if 0
+	if (!(mbuf->ol_flags & PKT_TX_IEEE1588_TMST))
+		return false;
+
+	type = ptp_classify_raw(mbuf);
+	if (!type)
+		return false;
+
+	hdr = ptp_parse_header(mbuf, type);
+	if (!hdr)
+		return false;
+
+	skb_shinfo(mbuf)->tx_flags |= SKBTX_IN_PROGRESS;
+#else
+  type = PTP_CLASS_V1;
+#endif
+
+	ptpdesc = MVPP22_PTP_MACTIMESTAMPINGEN |
+		  MVPP22_PTP_ACTION_CAPTURE;
+	queue = &priv->tx_hwtstamp_queue[0];
+
+	switch (type & PTP_CLASS_VMASK) {
+	case PTP_CLASS_V1:
+		ptpdesc |= MVPP22_PTP_PACKETFORMAT(MVPP22_PTP_PKT_FMT_PTPV1);
+		break;
+
+#if 0
+	case PTP_CLASS_V2:
+		ptpdesc |= MVPP22_PTP_PACKETFORMAT(MVPP22_PTP_PKT_FMT_PTPV2);
+		mtype = hdr->msg_type & 15;
+		/* Direct PTP Sync messages to queue 1 */
+		if (mtype == 0) {
+			ptpdesc |= MVPP22_PTP_TIMESTAMPQUEUESELECT;
+			queue = &priv->tx_hwtstamp_queue[1];
+		}
+		break;
+#endif
+	}
+
+	/* Take a reference on the mbuf and insert into our queue */
+	i = queue->next;
+	/* Iterate over 0..31 */
+	queue->next = (i + 1) & 31;
+	rte_pktmbuf_free(queue->mbuf[i]);
+	queue->mbuf[i] = rte_pktmbuf_clone(mbuf, mbuf->pool);
```

Comment:
why clone the buffer? (instead of adding a refcount)

Reply:
It's a proper way to increment refcount in DPDK

Reply:
You are right. The naming is just strange.